cpp/src/arrow/array/builder_union.h

```diff
@@ -54,53 +94,78 @@ class ARROW_EXPORT DenseUnionBuilder : public ArrayBuilder {
   }
 
   Status AppendNulls(int64_t length) final {
-    ARROW_RETURN_NOT_OK(types_builder_.Reserve(length));
-    ARROW_RETURN_NOT_OK(offsets_builder_.Reserve(length));
-    ARROW_RETURN_NOT_OK(Reserve(length));
-    for (int64_t i = 0; i < length; ++i) {
-      types_builder_.UnsafeAppend(0);
-      offsets_builder_.UnsafeAppend(0);
-    }
+    ARROW_RETURN_NOT_OK(types_builder_.Append(length, 0));
+    ARROW_RETURN_NOT_OK(offsets_builder_.Append(length, 0));
     return AppendToBitmap(length, false);
   }
 
   /// \brief Append an element to the UnionArray. This must be followed
   ///        by an append to the appropriate child builder.
-  /// \param[in] type index of the child the value will be appended
-  /// \param[in] offset offset of the value in that child
-  Status Append(int8_t type, int32_t offset) {
-    ARROW_RETURN_NOT_OK(types_builder_.Append(type));
+  ///
+  /// \param[in] next_type type_id of the child to which the next value will be appended.
+  ///
+  /// The corresponding child builder must be appended to independently after this method
+  /// is called.
+  Status Append(int8_t next_type) {
+    ARROW_RETURN_NOT_OK(types_builder_.Append(next_type));
+    if (type_id_to_children_[next_type]->length() == kListMaximumElements) {
```

Comment:
Do we want to defer this to Finish? We'd save a lot of checks for the "general" case.

Reply:
We'll get signed overflow in the offsets buffer if we don't do the check here. I can slap that into defined behavior with casts, though. Is that preferable?

Reply:
ListBuilder checks this on every append; I think I'd prefer to bail early if the array is going to be invalid later. Wontfix unless you insist

Reply:
:+1: